File: backend/app/schemas/analytics.py
"""
Analytics Schemas
Pydantic models for analytics API responses
"""
from pydantic import BaseModel


class ExecutionTrend(BaseModel):
    """Execution trend data for a specific date"""

    date: str
    total: int
    success: int
    failed: int


class StatusDistribution(BaseModel):
    """Distribution of execution statuses"""

    status: str
    count: int


class PipelinePerformance(BaseModel):
    """Performance metrics for a pipeline"""

    pipeline_id: str
    pipeline_name: str
    execution_count: int
    avg_duration_seconds: float
    success_rate: float


class AnalyticsOverview(BaseModel):
    """Overview of analytics data with KPIs and trends"""

    # KPIs
    total_executions: int
    successful_executions: int
    failed_executions: int
    running_executions: int
    success_rate: float  # percentage
    avg_duration_seconds: float

    # Trends and distributions
    execution_trends: list[ExecutionTrend]
    status_distribution: list[StatusDistribution]
    pipeline_performance: list[PipelinePerformance]

    # Metadata
    period_days: int

File: frontend/src/api/analytics.ts
/**
 * Analytics API Client
 * Handles requests to the analytics endpoints
 */
import { api } from './client';

// Types
export interface ExecutionTrend {
  date: string;
  total: number;
  success: number;
  failed: number;
}

export interface StatusDistribution {
  status: string;
  count: number;
}

export interface PipelinePerformance {
  pipeline_id: string;
  pipeline_name: string;
  execution_count: number;
  avg_duration_seconds: number;
  success_rate: number;
}

export interface AnalyticsOverview {
  // KPIs
  total_executions: number;
  successful_executions: number;
  failed_executions: number;
  running_executions: number;
  success_rate: number;
  avg_duration_seconds: number;

  // Trends and distributions
  execution_trends: ExecutionTrend[];
  status_distribution: StatusDistribution[];
  pipeline_performance: PipelinePerformance[];

  // Metadata
  period_days: number;
}

export interface AnalyticsParams {
  days?: number;
  pipeline_id?: string;
}

// API Functions
export const getAnalyticsOverview = async (
  params?: AnalyticsParams
): Promise<AnalyticsOverview> => {
  return api.get<AnalyticsOverview>('/api/v1/analytics/overview', { params });
};
